src/cargo/util/context/mod.rs

```diff
@@ -1537,36 +1537,32 @@ impl GlobalContext {
         let possible = dir.join(filename_without_extension);
         let possible_with_extension = dir.join(format!("{}.toml", filename_without_extension));
 
-        if possible.exists() {
+        if let Ok(possible_handle) = same_file::Handle::from_path(&possible) {
             if warn {
+                if let Ok(possible_with_extension_handle) =
+                    same_file::Handle::from_path(&possible_with_extension)
+                {
                 // We don't want to print a warning if the version
                 // without the extension is just a symlink to the version
                 // WITH an extension, which people may want to do to
                 // support multiple Cargo versions at once and not
                 // get a warning.
-                let skip_warning = if let Ok(target_path) = fs::read_link(&possible) {
-                    target_path == possible_with_extension
-                } else {
-                    false
-                };
-
-                if !skip_warning {
-                    if possible_with_extension.exists() {
+                    if possible_handle != possible_with_extension_handle {
```

Comment:
I did this deliberately.  I think it's fine for either side to be a symlink.  Let me explain:

The reason we are warning is as follows: if the user has both these files, and they are different, then different versions of cargo might behave differently due to reading different files.  That would be quite confusing.

If both files are the same, whichever way round they are symlinked, then this risk doesn't arise.

So insisting on the symlink being made in one particular direction would just be being pernickety for no good reason.  (And also, our current  warning message, inviting the user to make a symlink, is ambiguous, because the phrase "link A to B" might mean "make A -> B" or "make B -> A".)

If we *do* want to insist on the symlink being made in one particular direction (which I think we don't) then we certainly ought to have a bespoke error message for this situation.

We would also then have to decide what we think if *both* files are symlinks referring, ultimately, to the same file, somewhere else.  Currently we're happy with this, as I think we should be.

Reply:
A change in behavior like that should be called out explicitly in an easy-to-discover way in the PR description